### PR TITLE
Add `Vector256` support to the locator

### PR DESCRIPTION
**Pull request by mvadari**
### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

### CODE REVIEW

**Comment by pwang200** [comment]
Look good to me, but Olek should take a look.

File: src/xrpld/app/wasm/detail/HostFuncImpl.cpp [hunk 1]
```diff
@@ -290,24 +309,25 @@ WasmHostFunctionsImpl::getLedgerObjNestedField(
...
         return Unexpected(HostFunctionError::NO_ARRAY);
 
     auto const* field = ctx.tx.peekAtPField(fname);
     if (noField(field))
         return Unexpected(HostFunctionError::FIELD_NOT_FOUND);
 
-    if (field->getSType() != STI_ARRAY)
-        return Unexpected(HostFunctionError::NO_ARRAY);  // LCOV_EXCL_LINE
-    int32_t const sz = static_cast<STArray const*>(field)->size();
+    if (field->getSType() == STI_VECTOR256)
```
Comment (oleks-rip) [inline src/xrpld/app/wasm/detail/HostFuncImpl.cpp]: Please put 319-324 into helper function, and use it in other places

**Review by oleks-rip — CHANGES_REQUESTED** [review]
Please put 319-324 into helper function, and use it in other places

**Comment by oleks-rip** [comment]
I have concern about static variable, if someone later will call that  function 2 times it will overwrite the value without any notice. Try this [variants.patch](https://github.com/user-attachments/files/24461138/variants.patch) patch with variants instead

File: src/test/app/HostFuncImpl_test.cpp [hunk 2]
```diff
@@ -299,13 +299,21 @@ struct HostFuncImpl_test : public beast::unit_test::suite
         testcase("getTxField");
         using namespace test::jtx;
 
+        static std::string const credIdHex =
```
Comment (oleks-rip) [inline src/test/app/HostFuncImpl_test.cpp]: You don't need static here and below

File: src/test/app/HostFuncImpl_test.cpp [hunk 3]
```diff
@@ -537,6 +551,11 @@ struct HostFuncImpl_test : public beast::unit_test::suite
         Env env{*this};
         OpenView ov{*env.current()};
 
+        static std::string const credIdHex =
```
Comment (oleks-rip) [inline src/test/app/HostFuncImpl_test.cpp]: You don't need static here

File: src/test/app/HostFuncImpl_test.cpp (hunk 2 above)
Comment (mvadari) [reply src/test/app/HostFuncImpl_test.cpp]: Why not?

Comment (oleks-rip) [reply src/test/app/HostFuncImpl_test.cpp]: I don't see a reason why you need it. All the functions that require it are in the same scope and in the same thread, so `credIdHex` doesn't need life prolongation, like other variables

File: src/xrpld/app/wasm/detail/HostFuncImpl.cpp [hunk 4]
```diff
@@ -255,12 +266,8 @@ locateField(STObject const& obj, Slice const& locator)
             auto const* v = static_cast<STVector256 const*>(field);
             if (sfieldCode >= v->size())
                 return Unexpected(HostFunctionError::INDEX_OUT_OF_BOUNDS);
-            // STVector256::operator[] returns uint256&, not STBase*
-            // We need to wrap it in an STUInt256 to get an STBase pointer
-            // Use the value-only constructor to avoid field type mismatch
-            static thread_local STUInt256 tempUInt256;
-            tempUInt256 = STUInt256(v->operator[](sfieldCode));
-            field = &tempUInt256;
+            return std::variant<STBase const*, uint256 const*>(
```
Comment (oleks-rip) [inline src/xrpld/app/wasm/detail/HostFuncImpl.cpp]: return STBaseOrUInt256(&(v->operator[](sfieldCode)))

File: src/xrpld/app/wasm/detail/HostFuncImpl.cpp [hunk 5]
```diff
@@ -271,7 +278,7 @@ locateField(STObject const& obj, Slice const& locator)
             return Unexpected(HostFunctionError::FIELD_NOT_FOUND);
     }
 
-    return field;
+    return std::variant<STBase const*, uint256 const*>(field);
```
Comment (oleks-rip) [inline src/xrpld/app/wasm/detail/HostFuncImpl.cpp]: return STBaseOrUInt256(field)

File: src/xrpld/app/wasm/detail/HostFuncImpl.cpp [hunk 6]
```diff
@@ -381,7 +393,7 @@ WasmHostFunctionsImpl::getTxNestedArrayLen(Slice const& locator)
     if (!r)
         return Unexpected(r.error());
 
-    auto const* field = r.value();
+    auto const field = r.value();
```
Comment (oleks-rip) [inline src/xrpld/app/wasm/detail/HostFuncImpl.cpp]: `auto const &`

File: src/xrpld/app/wasm/detail/HostFuncImpl.cpp [hunk 7]
```diff
@@ -5,12 +5,16 @@
...
 
+#include <variant>
+
 #ifdef _DEBUG
 // #define DEBUG_OUTPUT 1
 #endif
 
 namespace xrpl {
 
+typedef std::variant<STBase const*, uint256 const*> STBaseOrUInt256;
```
Comment (oleks-rip) [inline src/xrpld/app/wasm/detail/HostFuncImpl.cpp]: The name should be more general, like `FieldValue` or something

File: src/test/app/HostFuncImpl_test.cpp [hunk 8]
```diff
@@ -336,6 +344,12 @@ struct HostFuncImpl_test : public beast::unit_test::suite
                 BEAST_EXPECT(
                     memos.error() == HostFunctionError::NOT_LEAF_FIELD);
 
+            auto const credentialIds = hfs.getTxField(sfCredentialIDs);
```
Comment (oleks-rip) [inline src/test/app/HostFuncImpl_test.cpp]: Better to leave comment here, because it is not clear thy it should return error, if one will see that field is present in the transaction

**Review by oleks-rip — CHANGES_REQUESTED** [review]
Please add test with maximum allowed size for STI_VECTOR256

**Comment by mvadari** [comment]
> Please add test with maximum allowed size for STI_VECTOR256

There's no maximum size, it depends on the field.

File: src/test/app/HostFuncImpl_test.cpp [hunk 9]
```diff
@@ -14,21 +14,27 @@ toBytes(std::uint8_t value)
 }
 
 static Bytes
-toBytes(std::uint16_t value)
+toBytes(std::uint16_t const value)
```
Comment (oleks-rip) [inline src/test/app/HostFuncImpl_test.cpp]: we don't put `const` for simple values

File: src/test/app/HostFuncImpl_test.cpp [hunk 10]
```diff
@@ -661,6 +721,10 @@ struct HostFuncImpl_test : public beast::unit_test::suite
         // Locator for STArray
         expectError({sfMemos.fieldCode}, HostFunctionError::NOT_LEAF_FIELD);
 
+        // Locator for STVector256
+        expectError(
```
Comment (oleks-rip) [inline src/test/app/HostFuncImpl_test.cpp:725]: this test duplicates one at 357 (imo this better)

Comment (mvadari) [reply src/test/app/HostFuncImpl_test.cpp:725]: Two different types of functions - one is nested, one is not

File: src/test/app/HostFuncImpl_test.cpp (hunk 9 above)
Comment (mvadari) [reply src/test/app/HostFuncImpl_test.cpp]: 1534559f648700c33cbe88377d766357a945f331

**Review by oleks-rip — APPROVED** [review]
